language/move-lang/tests/move_check/parser/spec_parsing_ok.move

```diff
@@ -0,0 +1,101 @@
+// These are some basic parsing tests for specifications which are expected to succeed.
+// Full testing of spec parsing correctness is done outside of this crate.
+//
+// Note that even though we only test parsing, we still need to ensure that the move code (not the specification)
+// is type checking correctly, because with no parsing error, the test harness
+// will run subsequent phases of the move-lang compiler.
+//
+// For parse failures, see the `spec_*_fail.move` test cases.
+
+module M {
+    synthetic expected_coin_sum: u64;
+    synthetic other: bool;
+
+    resource struct T{x: u64}
+    resource struct R{x: u64}
+
+    struct SomeCoin {
+        x : u64,
+        invariant x > 0,
+        y: u64,
+        invariant x == y,
+        invariant update old(y) < x,  // this weird coin only allows to increase it's value
+
+        // Tracking synthetic variable use case.
+        invariant update expected_coin_sum = expected_coin_sum - old(x) + x,
+        invariant pack expected_coin_sum = expected_coin_sum + x,
+        invariant unpack expected_coin_sum = expected_coin_sum - x,
+    }
+
+    fun with_aborts_if(x: u64)
+        aborts_if x == 0
+    {
+    }
+
+    fun with_ensures(x: u64): u64
+        ensures RET == x + 1
+    {
+        x + 1
+    }
+
+    fun with_multiple_conditions_and_acquires(addr: address, x: u64, y: u64): u64
+        acquires T, R
+        aborts_if y == 0
+        aborts_if 0 == y
+        ensures RET == x/y
+        ensures x/y == RET
+    {
+        let _ = borrow_global_mut<T>(addr);
+        let _ = borrow_global_mut<R>(addr);
+        x / y
+    }
+
+    fun using_pack(x: u64): T
+        ensures RET = T{x:x}  // spec expressions require no space after T for disambiguation with sequence start
+    {
+        T {x:x} // this does not apply to pack expressions outside of spec context
+    }
+
+    fun using_block(x: u64): u64
+        ensures RET = {let y = x; y + 1}
+    {
+        x + 1
+    }
+
+    fun using_lambda(x: u64): u64
+        ensures all(x, fun (y, z) x + y + z)
+    {
+        x
+    }
+
+    fun using_index_and_range(x: u64): u64
+        ensures RET = x[1] && x[0..3]
+    {
+        x
+    }
+
+    fun using_implies(x: u64): u64
+        ensures x > 0 ==> RET == x - 1
+        ensures x == 0 ==> RET == x
+    {
+        x
+    }
+
+    fun using_receiver_style_call(x: address)
+        aborts_if x.is_valid()
+        aborts_if x.is_valid_according_to_doug(42)
+        aborts_if x.is_valid_according_to_garp(17, 26)
+    {}
+
+    spec fun spec_fun_using_spec_constructs(x: u64, y: u64) {
```

Comment:
I'm a bit concerned about this

I almost want the syntax to feel more different, as the function is not being run through the same pass. 
In other words, the syntax looks similar but is treated WILDLY different by the compiler. If the compiler also did all of the typechecking and the like for these spec functions, it would be different. But as it stands, it feels unsettling. 

Maybe even drop the fun keyword?

Just 
```
spec spec_using_spec_constructs(x: u64, y: u64) {
}
```

Reply:
Functions are now in spec blocks